pkg/util/tolerations/tolerations.go

```diff
@@ -17,114 +17,91 @@ limitations under the License.
 package tolerations
 
 import (
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/klog"
 	api "k8s.io/kubernetes/pkg/apis/core"
 )
 
-type key struct {
-	tolerationKey string
-	effect        api.TaintEffect
-}
-
-func convertTolerationToKey(in api.Toleration) key {
-	return key{in.Key, in.Effect}
-}
-
 // VerifyAgainstWhitelist checks if the provided tolerations
 // satisfy the provided whitelist and returns true, otherwise returns false
-func VerifyAgainstWhitelist(tolerations []api.Toleration, whitelist []api.Toleration) bool {
-	if len(whitelist) == 0 {
+func VerifyAgainstWhitelist(tolerations, whitelist []api.Toleration) bool {
+	if len(whitelist) == 0 || len(tolerations) == 0 {
 		return true
 	}
 
-	t := ConvertTolerationToAMap(tolerations)
-	w := ConvertTolerationToAMap(whitelist)
-
-	for k1, v1 := range t {
-		if v2, ok := w[k1]; !ok || !AreEqual(v1, v2) {
-			return false
+next:
+	for _, t := range tolerations {
+		for _, w := range whitelist {
+			if isSuperset(w, t) {
+				continue next
+			}
 		}
+		return false
 	}
+
 	return true
 }
 
-// IsConflict returns true if the key of two tolerations match
-// but one or more other fields differ, otherwise returns false
-func IsConflict(first []api.Toleration, second []api.Toleration) bool {
-	firstMap := ConvertTolerationToAMap(first)
-	secondMap := ConvertTolerationToAMap(second)
-
-	for k1, v1 := range firstMap {
-		if v2, ok := secondMap[k1]; ok && !AreEqual(v1, v2) {
-			return true
+// MergeTolerations merges two sets of tolerations into one. If one toleration is a superset of
+// another, only the superset is kept.
+func MergeTolerations(first, second []api.Toleration) []api.Toleration {
+	all := append(first, second...)
+	merged := []api.Toleration{}
+
+next:
+	for i, t := range all {
+		for _, t2 := range merged {
+			if isSuperset(t2, t) {
+				continue next // t is redundant; ignore it
+			}
 		}
-	}
-	return false
-}
-
-// MergeTolerations merges two sets of tolerations into one
-// it does not check for conflicts
-func MergeTolerations(first []api.Toleration, second []api.Toleration) []api.Toleration {
-	var mergedTolerations []api.Toleration
-	mergedTolerations = append(mergedTolerations, second...)
-	firstMap := ConvertTolerationToAMap(first)
-	secondMap := ConvertTolerationToAMap(second)
-	// preserve order of first when merging
-	for _, v := range first {
-		k := convertTolerationToKey(v)
-		// if first contains key conflicts, the last one takes precedence
-		if _, ok := secondMap[k]; !ok && firstMap[k] == v {
-			mergedTolerations = append(mergedTolerations, v)
+		if i+1 < len(all) {
+			for _, t2 := range all[i+1:] {
+				// If the tolerations are equal, prefer the first.
+				if isSuperset(t2, t) && !apiequality.Semantic.DeepEqual(&t, &t2) {
```

Comment:
apiequality.Semantic.DeepEqual is called by isSuperset.

I think the order of the conditions should be :
```
				if !apiequality.Semantic.DeepEqual(&t, &t2) && isSuperset(t2, t) {
```
This way, if apiequality.Semantic.DeepEqual(&t, &t2) == true, isSuperset won't be called.